core/iwasm/compilation/aot_emit_aot_file.c

```diff
@@ -1506,31 +1506,143 @@ aot_emit_table_info(uint8 *buf, uint8 *buf_end, uint32 *p_offset,
+ * AOTCompData.
+ *
+ * @param comp_data The AOTCompData containing the type list.
+ * @param type The AOTType to search for.
+ *
+ * @return The index of the given type in the type list, or (uint32)-1 if not
+ * found.
+ */
+static uint32
+aot_emit_get_type_idx(AOTCompData *comp_data, AOTType *type)
+{
+    uint32 i;
+    for (i = 0; i < comp_data->type_count; i++) {
+        if (comp_data->types[i] == type)
+            return i;
+    }
+    return (uint32)-1;
+}
+
+static bool
+aot_emit_reftype_map(uint8 *buf, uint8 *buf_end, uint32 *p_offset, uint32 count,
+                     WASMRefTypeMap *refmap)
+{
+    uint32 offset = *p_offset, i;
+
+    for (i = 0; i < count; i++) {
+        EMIT_U16(refmap[i].index);
+    }
+
+    *p_offset = offset;
+    return true;
+}
+
 static bool
 aot_emit_type_info(uint8 *buf, uint8 *buf_end, uint32 *p_offset,
-                   AOTCompData *comp_data, AOTObjectData *obj_data)
+                   AOTCompContext *comp_ctx, AOTCompData *comp_data,
+                   AOTObjectData *obj_data)
 {
     uint32 offset = *p_offset, i;
-    AOTFuncType **func_types = (AOTFuncType **)comp_data->types;
+    int32 idx;
 
     *p_offset = offset = align_uint(offset, 4);
 
     EMIT_U32(comp_data->type_count);
 
-    for (i = 0; i < comp_data->type_count; i++) {
-        offset = align_uint(offset, 4);
-        EMIT_U32(func_types[i]->param_count);
-        EMIT_U32(func_types[i]->result_count);
-        EMIT_BUF(func_types[i]->types,
-                 func_types[i]->param_count + func_types[i]->result_count);
-    }
+    if (comp_ctx->enable_gc) {
+        AOTType **types = comp_data->types;
 
-    if (offset - *p_offset != get_func_type_info_size(comp_data)) {
-        aot_set_last_error("emit function type info failed.");
-        return false;
+        for (i = 0; i < comp_data->type_count; i++) {
+            offset = align_uint(offset, 4);
+            EMIT_U16(types[i]->type_flag);
+            EMIT_U16(types[i]->is_sub_final);
+            EMIT_U32(types[i]->inherit_depth);
+            /* root type index instead of type pointer */
+            idx = aot_emit_get_type_idx(comp_data, types[i]->root_type);
+            if (idx < 0) {
+                aot_set_last_error("invalid root type.");
+                return false;
+            }
+            EMIT_U32(idx);
+            EMIT_U32(types[i]->parent_type_idx);
+
+            /* Emit DEFINED_TYPE_FUNC */
+            if (types[i]->type_flag == DEFINED_TYPE_FUNC) {
+                AOTFuncType *func_type = (AOTFuncType *)types[i];
+                EMIT_U16(func_type->param_count);
+                EMIT_U16(func_type->result_count);
+                EMIT_U16(func_type->param_cell_num);
+                EMIT_U16(func_type->ret_cell_num);
+                EMIT_U16(func_type->ref_type_map_count);
+                EMIT_BUF(func_type->types,
+                         func_type->param_count * func_type->result_count);
+            }
```

Comment:
There are some misunderstandings, please refer to the new document.

Reply:
OK